### ISSUE TITLE
rasm2 does not load dynamic plugins

### ISSUE THREAD

**Issue by Lukas-Dresel** [radare/radare2#4495]
r2 correctly loads dynamic libraries found in $HOME/bin/prefix/radare2/lib/radare2/0.10.2-git, however rasm2 does not as can be seen by comparing `rasm2 -L` and running `e asm.arch=?` from inside r2.
Results:
[r2_asm_arch_list.txt](https://github.com/radare/radare2/files/201035/r2_asm_arch_list.txt)
[rasm2_list.txt](https://github.com/radare/radare2/files/201036/rasm2_list.txt)

The last 3 plugins (cil, mycpu, propeller) are not found by rasm2.

The mycpu plugin is from https://github.com/radare/radare2-extras/tree/master/examples/my-cpu
after running `make && make install`

r2 was installed for the current user using sys/user.sh

In r2 running `e arch.asm=mycpu` works and switches to mycpu disassembly.
Running `rasm2 -D -a mycpu 90` returns:
'rasm2: Unknown asm plugin 'mycpu''

Comparing running with LD_DEBUG=libs R_DEBUG=1 shows the following when running

`LD_DEBUG=libs R_DEBUG=1 r2 -c 'e asm.arch=?' /bin/ls 2>r2_ld_debug.txt`
[r2_ld_debug.txt](https://github.com/radare/radare2/files/201052/r2_ld_debug.txt)

`LD_DEBUG=libs R_DEBUG=1 rasm2 -L 2>rasm2_ld_debug.txt`
[rasm2_ld_debug.txt](https://github.com/radare/radare2/files/201053/rasm2_ld_debug.txt)

**Comment by radare** [comment]
it works for me with `bpf` for example. Where are those plugins installed when you type `make install`?

i have r2 installed system wide. maybe its an issue with the user install script.

**Comment by Lukas-Dresel** [comment]
make install does: `cp -f mycpu.so /home/<user>/bin/prefix/radare2//lib/radare2/0.10.2-git

For comparison, `r2 -hh` returns:
[r2_hh.txt](https://github.com/radare/radare2/files/201090/r2_hh.txt)

**Comment by radare** [comment]
Oh. Using r2pm -i bpf puts the plugin in the home directory which works. But not for this directory, do you have the 'last' symlink?

**Comment by radare** [comment]
Strace | grep opendir may help

**Comment by Lukas-Dresel** [comment]
where is the 'last' symlink supposed to be? 
I do have `/home/\<user\>/bin/prefix/radare2/lib/radare2/last`

**Comment by Lukas-Dresel** [comment]
opendir is never called in any of the processes. 
I ran it with `strace -f -ff -o log.txt rasm2 -D -a mycpu 90`.
If you tell me exactly which process you want I can upload the corresponding log.

**Comment by Lukas-Dresel** [comment]
Here are all the process logs:

[strace_log.1427.txt](https://github.com/radare/radare2/files/201102/strace_log.1427.txt)
[strace_log.1429.txt](https://github.com/radare/radare2/files/201106/strace_log.1429.txt)
[strace_log.1430.txt](https://github.com/radare/radare2/files/201105/strace_log.1430.txt)
[strace_log.1431.txt](https://github.com/radare/radare2/files/201108/strace_log.1431.txt)
[strace_log.1432.txt](https://github.com/radare/radare2/files/201104/strace_log.1432.txt)
[strace_log.1433.txt](https://github.com/radare/radare2/files/201107/strace_log.1433.txt)
[strace_log.1434.txt](https://github.com/radare/radare2/files/201103/strace_log.1434.txt)
[strace_log.1435.txt](https://github.com/radare/radare2/files/201109/strace_log.1435.txt)
[strace_log.1436.txt](https://github.com/radare/radare2/files/201111/strace_log.1436.txt)
[strace_log.1437.txt](https://github.com/radare/radare2/files/201110/strace_log.1437.txt)
[strace_log.txt](https://github.com/radare/radare2/files/201112/strace_log.txt)

**Comment by Lukas-Dresel** [comment]
Okay, i looked some more into it, the important file is `strace_log.1437.txt`
From line 170-176 we can see that rasm2 tries to access the plugins from 
`/home/<user>/.config/radare2/plugins` while everything else seems to be correctly accessed via `/home/<user>/bin/prefix/radare2//lib/`.

I'm currently looking to find the problem, any ideas are appreciated

**Comment by Lukas-Dresel** [comment]
The bug seems to me to be in `binr/rasm2/rasm2.c` line 360. It hardcodes the plugin directory as in the home directory.

**Comment by Lukas-Dresel** [comment]
binr/radare2/radare2.c calls r_core_loadlibs() with R_CORE_LOADLIBS_ALL

instead rasm2 calls r_lib_opendir() with `R2_HOMEDIR "/plugins"`

**Comment by Lukas-Dresel** [comment]
r_core_loadlibs
checks in order:
path, configuration, environment, home directory, system libs

**Comment by Lukas-Dresel** [comment]
It seems to me like you would have to replace lines 346-365 with a corresponding r_core_loadlibs call.
As i'm not very familiar with the internals of r2 can anyone take a look at this? I'll attempt a fix but don't wanna break other stuff.

**Comment by radare** [comment]
i dont see which dir is missing in your case. the thing works fine for me. Just to clarify:

- it is just an issue when installing r2 in your home?
- the plugin is installed in your homedir or in “r2 system directory”
- which path is exposed by r2 -hh| grep PLUGINS

The patch should be just adding the missing r_lib_opendir call  in binr/rasm2/rasm2.c around line 346.

Can you give it a try?

**Comment by Lukas-Dresel** [comment]
Wouldn't you wanna have exactly the same code as in /libr/core/libs.c? for r_core_loadlibs function?
I get that we could fix this easily by just modifying it in rasm2 for now, but that would just pop up again if anything changes. How about this, i'll try fixing it and you decide on the pull request if you like it or not. If not i'll simply fix it in rasm2

**Comment by Lukas-Dresel** [comment]
Oh btw, to clarify: 
The problem is that rasm2 does not use LIBR_PLUGINS.
It's not a user vs system install problem, the problem seems to be that you guys mostly test with r2pm which according to you copies the plugins to the home directory. That's where rasm2 looks. The problem is, however, that if you use the makefiles they use LIBR_PLUGINS from `r2 -hh`. Rasm2 doesn't look there.
1. No, the issue is where the plugin is installed
2. the plugin is installed in LIBR_PLUGINS
3. I posted my `r2 -hh` logs above

I made a patch, pull-request will come as soon as i verify it still works ;)
If you wanna see it before-hand, look in my fork in branch lib_loading_fix
